zuul-core/src/main/java/com/netflix/zuul/netty/server/Server.java

```diff
@@ -57,67 +71,94 @@
+    public Server(Map<Integer, ChannelInitializer> portsToChannelInitializers, ServerStatusManager serverStatusManager,
+                  ClientConnectionsShutdown clientConnectionsShutdown, EventLoopGroupMetrics eventLoopGroupMetrics)
     {
-        this(portsToChannelInitializers, serverStatusManager, clientConnectionsShutdown, eventLoopGroupMetrics, new DefaultEventLoopConfig());
+        this(portsToChannelInitializers, serverStatusManager, clientConnectionsShutdown, eventLoopGroupMetrics,
+                new DefaultEventLoopConfig());
     }
 
-    public Server(Map<Integer, ChannelInitializer> portsToChannelInitializers, ServerStatusManager serverStatusManager, ClientConnectionsShutdown clientConnectionsShutdown, EventLoopGroupMetrics eventLoopGroupMetrics, EventLoopConfig eventLoopConfig)
-    {
-        this.portsToChannelInitializers = portsToChannelInitializers;
-        this.serverStatusManager = serverStatusManager;
-        this.clientConnectionsShutdown = clientConnectionsShutdown;
-        this.eventLoopConfig = eventLoopConfig;
-        this.eventLoopGroupMetrics = eventLoopGroupMetrics;
-        this.jvmShutdownHook = new Thread(() -> stop(), "Zuul-JVM-shutdown-hook");
+    @SuppressWarnings("unchecked") // Channel init map has the wrong generics and we can't fix without api breakage.
+    public Server(Map<Integer, ChannelInitializer> portsToChannelInitializers, ServerStatusManager serverStatusManager,
+                  ClientConnectionsShutdown clientConnectionsShutdown, EventLoopGroupMetrics eventLoopGroupMetrics,
+                  EventLoopConfig eventLoopConfig) {
+        this(serverStatusManager,
+                convertPortMap((Map<Integer, ChannelInitializer<?>>) (Map) portsToChannelInitializers),
+                clientConnectionsShutdown, eventLoopGroupMetrics, eventLoopConfig);
+    }
+
+    Server(ServerStatusManager serverStatusManager,
+           Map<? extends SocketAddress, ? extends ChannelInitializer<?>> addressesToInitializers,
+           ClientConnectionsShutdown clientConnectionsShutdown, EventLoopGroupMetrics eventLoopGroupMetrics,
+           EventLoopConfig eventLoopConfig) {
+        this.addressesToInitializers = Collections.unmodifiableMap(new HashMap<>(addressesToInitializers));
+        this.serverStatusManager = checkNotNull(serverStatusManager, "serverStatusManager");
+        this.clientConnectionsShutdown = checkNotNull(clientConnectionsShutdown, "clientConnectionsShutdown");
+        this.eventLoopConfig = checkNotNull(eventLoopConfig, "eventLoopConfig");
+        this.eventLoopGroupMetrics = checkNotNull(eventLoopGroupMetrics, "eventLoopGroupMetrics");
     }
 
     public void stop()
     {
-        LOG.warn("Shutting down Zuul.");
+        LOG.debug("Shutting down Zuul.");
```

Comment:
Why change these to debug?

Reply:
As mentioned above, they aren't actionable.  Also, it doesn't much matter in the case of JVM shutdown, since the logger is usually gone by that point too.   

I posit that if someone was interested in the shutdown status, we should expose this via a Future, CountDownLatch, or something like that.   The reason is that callers may have their own logging preferences.

Reply:
upgraded to info, but I think these should still be logged at a different level